sklearn/cluster/_feature_agglomeration.py

```diff
@@ -40,14 +41,21 @@ def transform(self, X):
 
         pooling_func = self.pooling_func
         X = check_array(X)
-        nX = []
         if len(self.labels_) != X.shape[1]:
             raise ValueError("X has a different number of features than "
                              "during fitting.")
-
-        for l in np.unique(self.labels_):
-            nX.append(pooling_func(X[:, self.labels_ == l], axis=1))
-        return np.array(nX).T
+        if pooling_func == np.mean:
+            size = np.bincount(self.labels_)
+            n_samples = X.shape[0]
+            # a fast way to compute the mean of grouped features
+            nX = np.array([np.bincount(self.labels_, X[i, :])/size
```

Comment:
how does this weighted bincount compare to
```
X @ scipy.sparse.csr_matrix((np.ones_like(labels), labels, np.arange(len(labels) + 1)))
```

Reply:
this did not perform better. please see the plots below.